File: Top Interview Questions/Easy Collection/Dynamic Programming/Maximum Subarray.py
"""
URL : https://www.cnblogs.com/zuoyuan/p/3781988.html
"""
import sys

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        currentSum = 0
        maxSum = -sys.maxsize - 1

        for i in range(len(nums)):
            if currentSum < 0:
                currentSum = 0

            currentSum += nums[i]
            maxSum = max(currentSum, maxSum)

        return maxSum

File: Top Interview Questions/Easy Collection/Dynamic Programming/Maximum Subarray.java
// REF: https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm
class Solution {
    public int maxSubArray(int[] nums) {
        int currentSubarray = nums[0];
        int maxSubarray = nums[0];

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            currentSubarray = Math.max(num, currentSubarray + num);
            maxSubarray = Math.max(maxSubarray, currentSubarray);
        }

        return maxSubarray;
    }
}